### DISCUSSION TITLE
Does enabling ECC on the TCM correct single bit errors in the actual memory?

### DISCUSSION THREAD

**Discussion by andrenemat** [Mi-V-Soft-RISC-V/miv-rv32-documentation#13] [Mi-V RV32 Soft Processors] [answered]
Section 4.14 of the handbook goes into some details of the operation of the ECC. When ECC is enabled on the TCM, the CPU doesn't see the error, but does it actually correct and overwrite a single bit error in the physical memory, or would this need to be handled in software in the ECC interrupt. If it needs to be handled in software, do we an example how to handle this?

Can the TAS be accessed when the CPU is halted, or inly while it is in reset?

**Comment by hughbreslin** [comment, marked as answer]
Hey @andrenemat 

the ECC on the TCM doesn't perform writeback of corrected data, it just corrects the data for the read but the "bad" data will still be in memory. There is example code in section 4.14 of the handbook to scrub memory, this will read and write back data to memory so any corrected data will be written back.

```C
volatile uint32_t *tcm_addr;
tcm_addr = 0x40000000; // Memory mapped start address of TCM
uint32_t tcm_end_addr = 0x40002000; // Memory mapped end address of TCM
while (tcm_addr != tcm_end_addr){
*tcm_addr = *tcm_addr; 
tcm_addr = tcm_addr + 0x1;
}
```

You will need to match the tcm_addr and tcm_end_addr values to match your configuration or take them from the linker. 

It is up to you how this could be implemented, for example this code could be added to the single bit correct interrupt handler.

The TAS should be accessible in any stage (it won't be usable while the core is in reset) but it is recommended to only access it while the core is halted and you are writing data to prevent a race condition.

**Reply by andrenemat** [reply]
Thanks @hughbreslin . The reason for the question about the TAS was the section below from the handbook which implies the TAS is used when the processor is in reset.

2.8 TAS Port
The TAS port allows reading and writing to the TCM from an external source before the core is brought out of reset. It
is recommended that the address widths for the TCM and the TAS are of the same size to avoid memory read or
write violations.

**Reply by hughbreslin** [reply]
Hmm ok I’ll double check this, as far as I remember you could try to access the tcm but as it’s a shared reset for the hart and memory I would imagine the accesses will hang

**Reply by hughbreslin** [reply]
We chatted this over internally, my original statement is accurate whereby you can't access the TCM via the TAS while the core is in reset - you can access at any point when the core is out of reset and then you can also use the "disable access" signals to block an initiator from accessing the TCM at any point. We'll update the handbook to remove the reset statement and clarify the wording.

**Comment by andrenemat** [comment]
@hughbreslin Thank you very much for the clarification, appreciate all your help and advice
